tools/x-repo-types/typeAnalyzer/typeAnalyzer.go

```diff
@@ -0,0 +1,575 @@
+	for i, node := range t {
+		parts[i] = fmt.Sprintf("%q", node.String())
+	}
+	return fmt.Sprintf("[]string{%s}", strings.Join(parts, ", "))
+}
+
+// IsLeaf returns true if the TypeNode has no children.
+func (t *TypeNode) IsLeaf() bool {
+	return t.children == nil || len(*t.children) == 0
+}
+
+// Build the TypeNode tree by finding all child types that belong to a kind and recursively
+// building their TypeNode trees. Stop traversing when a cycle is detected.
+func (t *TypeNode) Build() TypePath {
+	return t.build(TypePath{})
+}
+
+func (t *TypeNode) appendChild(typeName, typeTag string, child TypeNode) {
+	cname := ChildName{typeName, typeTag}
+	t.ChildNames = append(t.ChildNames, cname)
+	if t.children == nil {
+		children := make(ChildTypes)
+		t.children = &children
+	}
+	(*t.children)[cname.String()] = child
+}
+
+// build the TypeNode tree by finding all child types that belong to a kind and recursively
+// building their TypeNode trees.
+// Return a path that will be non-trivial only in a case that a cycle is detected.
+func (t *TypeNode) build(path TypePath) TypePath {
+	if t.IsStruct() {
+		me := t.String()
+		foundCycle := false
+		for _, node := range path {
+			if node.String() == me {
+				foundCycle = true
+				break
+			}
+		}
+		path = append(path, *t)
+		if foundCycle {
+			return path
+		}
+	}
+
+	var cyclePath TypePath
+	switch t.Kind {
+	case reflect.Struct:
+		cyclePath = t.buildStructChildren(path)
+	case reflect.Slice, reflect.Array:
+		cyclePath = t.buildListChild(path)
+	case reflect.Map:
+		cyclePath = t.buildMapChildren(path)
+	case reflect.Ptr:
+		cyclePath = t.buildPtrChild(path)
+	}
+
+	return cyclePath
+}
+
+// buildStructChildren builds the children of a struct type.
+func (t *TypeNode) buildStructChildren(path TypePath) TypePath {
+	var cyclePath TypePath
+	for i := 0; i < t.Type.NumField(); i++ {
+		typeField := t.Type.Field(i)
+		typeName := typeField.Name
+
+		// probably we need to skip typeField.Tag == `codec:"-"` as well
+		if typeName == "" || (!unicode.IsUpper(rune(typeName[0])) && typeName != "_struct") {
+			continue
+		}
+
+		if typeField.Anonymous {
+			// embedded struct case
+			actualKind := typeField.Type.Kind()
+			if actualKind != reflect.Struct {
+				panic(fmt.Sprintf("expected [%s] but got unexpected embedded type: %s", reflect.Struct, typeField.Type))
+			}
+
+			embedded := TypeNode{t.Depth, typeField.Type, reflect.Struct, nil, nil}
+			embeddedCyclePath := embedded.build(path)
+			if len(embeddedCyclePath) > 0 {
+				cyclePath = embeddedCyclePath
+			}
+			for _, edge := range embedded.ChildNames {
+				child := (*embedded.children)[edge.String()]
+				t.appendChild(edge.Name, edge.Tag, child)
+			}
+			continue
+		}
+
+		typeTag := string(typeField.Tag)
+		child := TypeNode{t.Depth + 1, typeField.Type, typeField.Type.Kind(), nil, nil}
+		childCyclePath := child.build(path)
+		if len(childCyclePath) > 0 {
+			cyclePath = childCyclePath
+		}
+		t.appendChild(typeName, typeTag, child)
```

Comment:
If I'm reading correctly, like the only difference in the embedded case and the normal case is:
* An additional error check: `typeField.Type.Kind()` must be struct (therefore the hard coded `reflect.Struct` is the same as the code below.
* The depth is not incremented.
* When the child is appended, the parent is or is not included.

Could you add a comment to this effect? It took me a while to notice these differences since there were a lot of similarities.

Also, does this code support adjacent embedded types? I think the answer is yes, because build is called recursively in both cases.

Reply:
> Also, does this code support adjacent embedded types? I think the answer is yes, because build is called recursively in both cases.

Yes, in the for-loop it will pick up all fields, embedded (i.e. _Anonymous_) and not.

> If I'm reading correctly...
I think you're reading correctly. The main difference is not adding to the depth so that a compared struct would still be considered equal if it inlined all the embedded struct's fields (depths are required to be the same in the comparison algorithm).
I'll have a close look and add better comments.

Reply:
I added a test case for multiple levels of embedded structs. Worked perfectly.

Reply:
122e0ffc7e4949a9f5a659c2f9b61fef7e167a01